File: frontend/lib/operations.ts
/**
 * Reading the collector operations report: every decision the status page makes, as pure
 * functions.
 *
 * The page itself renders; it decides nothing. That split matters more here than elsewhere,
 * because this screen is the one an operator consults when something is wrong, and the
 * wording of "wrong" is the product. A sentence that says "failed" where it should say
 * "succeeded, but three days ago" sends somebody to fix a collector that is working.
 *
 * Nothing here re-derives a verdict. `completeness`, `note`, `shortfall`, `stale_success`
 * and `health` all come from the API, which computes them in
 * `app/domain/operations.py`. What this module adds is presentation: tone, ordering,
 * grouping and the two judgements that are genuinely about the screen — which counts are
 * worth calling out as zero, and how to word a timestamp that is old.
 */

import type {
  CollectionOperations,
  CollectorErrorGroup,
  Completeness,
  ObjectCounts,
  RunOutcome,
  ScopeOperations,
} from "@/lib/contracts";

/** How a completeness verdict should read, and how loudly. */
export interface Tone {
  /** The CSS status class, matching the rest of the application. */
  className: "status-ok" | "status-warn" | "status-bad" | "muted";
  label: string;
}

const TONES: Record<Completeness, Tone> = {
  complete: { className: "status-ok", label: "complete" },
  partial: { className: "status-warn", label: "incomplete" },
  none: { className: "status-bad", label: "nothing collected" },
  in_progress: { className: "muted", label: "running" },
};

export function completenessTone(value: Completeness): Tone {
  return TONES[value];
}

/**
 * Timestamps are rendered in UTC, always.
 *
 * A scan run's time is compared against a Windows event log and a collector log, both of
 * which an operator reads in a fixed zone. Rendering in the browser's zone would also make
 * the server and client markup disagree and produce a hydration mismatch.
 */
export function formatUtc(value: string | null): string {
  if (value === null) {
    return "—";
  }
  const parsed = new Date(value);
  if (Number.isNaN(parsed.getTime())) {
    return value;
  }
  return `${parsed.toISOString().replace("T", " ").slice(0, 19)} UTC`;
}

/**
 * How old a run is, in words, relative to a reference instant.
 *
 * The reference is a parameter rather than `Date.now()` so that this is testable and so
 * that the page passes one instant to every row: two rows computed from two different
 * "now"s can disagree by a minute and look like a bug.
 *
 * Returns `null` for anything under an hour — an operator does not need to be told that a
 * run from twenty minutes ago is recent, and a line that says so is a line in the way.
 */
export function ageInWords(value: string | null, now: Date): string | null {
  if (value === null) {
    return null;
  }
  const parsed = new Date(value);
  if (Number.isNaN(parsed.getTime())) {
    return null;
  }
  const hours = Math.floor((now.getTime() - parsed.getTime()) / 3_600_000);
  if (hours < 1) {
    return null;
  }
  if (hours < 48) {
    return `${hours} hour${hours === 1 ? "" : "s"} ago`;
  }
  const days = Math.floor(hours / 24);
  return `${days} day${days === 1 ? "" : "s"} ago`;
}

/**
 * The sentence under a scope's name.
 *
 * The API's own `note` is used verbatim when it has one — it is the authoritative wording
 * and it already distinguishes "never ran" from "failed, and here is how old your data is".
 * The only case this adds is the healthy one, where the API says nothing because there is
 * nothing wrong, and a blank cell reads as missing rather than fine.
 */
export function scopeSentence(scope: ScopeOperations, now: Date): string {
  if (scope.note !== null) {
    return scope.note;
  }
  const age = ageInWords(scope.latest.completed_at ?? scope.latest.started_at, now);
  const when = age === null ? "just now" : age;
  return `Last read ${when}, with nothing reported missing.`;
}

/** Scope counts by completeness, for the one-line header. */
export interface ScopeTally {
  complete: number;
  partial: number;
  none: number;
  in_progress: number;
  total: number;
  neverSucceeded: number;
}

export function tallyScopes(scopes: readonly ScopeOperations[]): ScopeTally {
  const tally: ScopeTally = {
    complete: 0,
    partial: 0,
    none: 0,
    in_progress: 0,
    total: scopes.length,
    neverSucceeded: 0,
  };
  for (const scope of scopes) {
    tally[scope.completeness] += 1;
    if (!scope.has_ever_succeeded) {
      tally.neverSucceeded += 1;
    }
  }
  return tally;
}

/**
 * The header sentence: how much of the estate is actually readable.
 *
 * Phrased as a fraction of scopes rather than as a health word, because "3 of 8 scopes are
 * incomplete" is actionable and "incomplete" on its own is not.
 */
export function coverageSentence(tally: ScopeTally): string {
  if (tally.total === 0) {
    return "No collector has ever reported. Every list in ADG is empty because nothing ran.";
  }
  if (tally.complete === tally.total) {
    return `All ${tally.total} collection scopes are current and complete.`;
  }
  const parts: string[] = [`${tally.complete} of ${tally.total} scopes are complete`];
  if (tally.partial) {
    parts.push(`${tally.partial} incomplete`);
  }
  if (tally.none) {
    parts.push(`${tally.none} with nothing usable`);
  }
  if (tally.in_progress) {
    parts.push(`${tally.in_progress} still running`);
  }
  const never =
    tally.neverSucceeded > 0
      ? ` ${tally.neverSucceeded} scope(s) have never completed a run at all.`
      : "";
  return `${parts.join(", ")}.${never}`;
}

/** One row of the object-count panel. */
export interface CountRow {
  key: keyof ObjectCounts;
  label: string;
  value: number;
  /** Set when a zero here means a collector did not do its job. */
  concern: string | null;
}

const COUNT_LABELS: ReadonlyArray<{ key: keyof ObjectCounts; label: string; empty: string }> = [
  {
    key: "principals",
    label: "Principals",
    empty: "No identity has been collected, so no ACL entry can be attributed to anybody.",
  },
  {
    key: "membership_edges",
    label: "Memberships",
    empty: "No membership is known, so every group looks empty and access is under-reported.",
  },
  { key: "servers", label: "Servers", empty: "No server has been collected." },
  { key: "shares", label: "Shares", empty: "No share has been collected." },
  {
    key: "share_aces",
    label: "Share ACL entries",
    empty: "No share ACL was read; the share layer cannot limit anything in an answer.",
  },
  {
    key: "directories",
    label: "Directories",
    empty: "No file-system descriptor was read, so no effective access can be computed.",
  },
  {
    key: "ntfs_aces",
    label: "NTFS ACL entries",
    empty: "No NTFS entry was read, so every directory looks like it grants nobody access.",
  },
  { key: "scan_runs", label: "Scan runs", empty: "Nothing has ever run." },
];

/**
 * The counts, with a sentence attached to each zero.
 *
 * A zero is the number this panel exists for. A collector that ran cleanly and wrote no
 * rows reports `succeeded` and looks fine everywhere else in the product; the only place
 * that shows it is a count of what actually landed.
 */
export function countRows(counts: ObjectCounts): CountRow[] {
  return COUNT_LABELS.map(({ key, label, empty }) => ({
    key,
    label,
    value: counts[key],
    concern: counts[key] === 0 ? empty : null,
  }));
}

/** Error groups, with the widespread ones first: they are the ones that are not one object. */
export function orderedErrors(errors: readonly CollectorErrorGroup[]): CollectorErrorGroup[] {
  return [...errors].sort((left, right) => {
    if (left.is_widespread !== right.is_widespread) {
      return left.is_widespread ? -1 : 1;
    }
    if (left.count !== right.count) {
      return right.count - left.count;
    }
    return left.code.localeCompare(right.code);
  });
}

/**
 * What a run delivered against what it claimed, as a short string.
 *
 * Both halves are shown even when they agree. "88" alone invites the reader to assume it is
 * what was asked for; "88 of 88" says so.
 */
export function delivery(run: RunOutcome): string {
  const observations =
    run.observations_reported === null
      ? `${run.observations_applied}`
      : `${run.observations_applied} of ${run.observations_reported}`;
  const batches =
    run.batches_reported === null
      ? `${run.batches_received}`
      : `${run.batches_received} of ${run.batches_reported}`;
  return `${observations} observations in ${batches} batches`;
}

/**
 * Whether this report should be read as an alert rather than as a status.
 *
 * Driven by the API's `health`, not by counting rows here: the coverage banner every other
 * page shows is computed from the same field, and two screens disagreeing about whether the
 * estate is trustworthy is worse than either being wrong.
 */
export function isAlarming(report: CollectionOperations): boolean {
  return report.health === "failed" || report.health === "incomplete";
}

File: backend/app/domain/operations.py
"""The operator's view of collection: what ran, what it missed, and what is in the store.

:mod:`app.domain.collection` answers one question — *can an empty page be believed?* — and
answers it for everybody, on every screen. This module answers the operator's questions,
which are different and more specific:

* When did this scope last **succeed**? When did it last **fail**? Those are two facts, and
  reducing them to "the latest run" — which is what the coverage banner does — throws away
  the one an operator needs most. A scope whose latest run succeeded but which failed twice
  yesterday is not the same as one that has never failed.
* Was the run **complete**? A run can succeed and still not deliver everything it collected:
  it may have sent more batches than arrived, reported more observations than were applied,
  or declared a scope it never reconciled. Each of those is a silent shortfall.
* **How much is actually stored?** A collector that ran cleanly and wrote four rows did not
  do what it was configured to do, and nothing about its own status says so.
* **What went wrong, across everything?** Twelve `access_denied` on one share is a
  permissions problem; twelve spread over twelve servers is a service-account problem.

Everything here is pure: it folds records into a report and holds no query, no session and
no request. :mod:`app.repositories.operations` supplies the records.
"""

from __future__ import annotations

import datetime as dt
from collections.abc import Iterable, Mapping
from dataclasses import dataclass
from enum import StrEnum

from app.domain.collection import CollectionStatus
from app.domain.observation import ScanStatus

__all__ = [
    "Completeness",
    "ErrorGroup",
    "ObjectCounts",
    "OperationsReport",
    "RunOutcome",
    "ScopeOperations",
    "assemble_operations",
]


class Completeness(StrEnum):
    """How much of what a scope's latest run was asked to do actually landed."""

    COMPLETE = "complete"
    """Succeeded, no errors, every batch delivered, every declared scope reconciled."""

    PARTIAL = "partial"
    """It ran and something is missing: errors, an undelivered batch, or an unreconciled
    scope. What was collected is trustworthy; what is absent proves nothing."""

    NONE = "none"
    """The run failed or was canceled. Nothing about this scope can be concluded."""

    IN_PROGRESS = "in_progress"
    """Still running, or never closed. The numbers below it are a partial tally."""


@dataclass(frozen=True, slots=True)
class RunOutcome:
    """One run, reduced to what an operator judges it by."""

    run_id: str
    collector: str
    collector_host: str
    target: str | None
    status: ScanStatus
    started_at: dt.datetime
    completed_at: dt.datetime | None
    error_count: int
    batches_reported: int | None
    batches_received: int
    observations_reported: int | None
    observations_applied: int
    declared_scopes: int
    reconciled_scopes: int
    incremental: bool = False
    downgrade_reason: str | None = None

    @property
    def batches_missing(self) -> int:
        """Batches the collector says it sent that never arrived.

        ``batches_reported`` is None for a run that never closed, and an open run has not
        claimed a total yet — so nothing is missing, it is merely not finished.
        """
        if self.batches_reported is None:
            return 0
        return max(0, self.batches_reported - self.batches_received)

    @property
    def observations_missing(self) -> int:
        """Observations claimed but not stored.

        Non-zero without ``batches_missing`` means the batches arrived and carried fewer
        rows than the completion claimed — a collector-side counting bug, and worth
        surfacing rather than rounding away.
        """
        if self.observations_reported is None:
            return 0
        return max(0, self.observations_reported - self.observations_applied)

    @property
    def scopes_unreconciled(self) -> int:
        """Scopes a full run declared and never confirmed it had fully enumerated.

        Always zero for an **incremental** run, which by definition looks at part of its
        scope and is not expected to reconcile one. Counting those would mark every
        incremental run incomplete and teach an operator to ignore the column.
        """
        if self.incremental:
            return 0
        return max(0, self.declared_scopes - self.reconciled_scopes)

    @property
    def completeness(self) -> Completeness:
        if self.status in (ScanStatus.RUNNING, ScanStatus.PENDING):
            return Completeness.IN_PROGRESS
        if self.status in (ScanStatus.FAILED, ScanStatus.CANCELED):
            return Completeness.NONE
        if (
            self.status is ScanStatus.SUCCEEDED
            and not self.error_count
            and not self.batches_missing
            and not self.observations_missing
            and not self.scopes_unreconciled
        ):
            return Completeness.COMPLETE
        return Completeness.PARTIAL

    @property
    def shortfall(self) -> str | None:
        """What is missing from this run, in one sentence, or ``None``.

        Deliberately separate from :attr:`app.domain.collection.CollectorCoverage.concern`:
        that one warns a reader that a page may under-report, this one tells an operator
        what to go and fix.
        """
        parts: list[str] = []
        if self.batches_missing:
            parts.append(f"{self.batches_missing} batch(es) never arrived")
        if self.observations_missing:
            parts.append(f"{self.observations_missing} observation(s) were claimed but not stored")
        if self.scopes_unreconciled:
            parts.append(f"{self.scopes_unreconciled} declared scope(s) were not fully enumerated")
        if self.error_count:
            parts.append(f"{self.error_count} object(s) could not be read")
        if not parts:
            return None
        return "; ".join(parts) + "."


@dataclass(frozen=True, slots=True)
class ScopeOperations:
    """One collector against one target: the latest run, and the last of each outcome."""

    collector: str
    target: str | None
    latest: RunOutcome
    last_success: RunOutcome | None
    last_failure: RunOutcome | None

    @property
    def scope_label(self) -> str:
        return self.collector if self.target is None else f"{self.collector} ({self.target})"

    @property
    def completeness(self) -> Completeness:
        return self.latest.completeness

    @property
    def has_ever_succeeded(self) -> bool:
        return self.last_success is not None

    @property
    def stale_success(self) -> bool:
        """The last success is older than the latest run: the newest attempt did not work.

        This is the state a status page most often gets wrong. Showing only the latest run
        says "failed" and hides that yesterday's data is still there; showing only the last
        success says "succeeded" and hides that it is stale. Both are shown.
        """
        if self.last_success is None:
            return False
        return self.last_success.run_id != self.latest.run_id

    @property
    def note(self) -> str | None:
        """One sentence for the operator, or ``None`` when the scope is healthy."""
        if self.latest.completeness is Completeness.NONE:
            if self.last_success is None:
                return (
                    f"{self.scope_label} has never completed a run. Nothing about this "
                    "scope has ever been collected."
                )
            return (
                f"The latest {self.scope_label} run failed. The data on screen is from "
                f"{_moment(self.last_success.completed_at or self.last_success.started_at)} "
                "and has not been refreshed since."
            )
        if self.latest.completeness is Completeness.IN_PROGRESS:
            return f"A {self.scope_label} run is still open; its totals are not final."
        if self.latest.completeness is Completeness.PARTIAL:
            return f"The latest {self.scope_label} run was incomplete: {self.latest.shortfall}"
        return None


@dataclass(frozen=True, slots=True)
class ErrorGroup:
    """Every reported failure sharing one code, folded together."""

    code: str
    count: int
    collectors: tuple[str, ...]
    latest_occurred_at: dt.datetime | None
    sample_targets: tuple[str, ...]

    @property
    def is_widespread(self) -> bool:
        """Reported by more than one collector, so it is unlikely to be one bad object."""
        return len(self.collectors) > 1


@dataclass(frozen=True, slots=True)
class ObjectCounts:
    """What the store actually holds. A clean run that wrote nothing is still a problem."""

    principals: int = 0
    membership_edges: int = 0
    servers: int = 0
    shares: int = 0
    share_aces: int = 0
    directories: int = 0
    ntfs_aces: int = 0
    scan_runs: int = 0

    @property
    def total(self) -> int:
        """Collected objects. ``scan_runs`` is excluded: a run is the act, not the estate."""
        return (
            self.principals
            + self.membership_edges
            + self.servers
            + self.shares
            + self.share_aces
            + self.directories
            + self.ntfs_aces
        )

    def as_mapping(self) -> Mapping[str, int]:
        return {
            "principals": self.principals,
            "membership_edges": self.membership_edges,
            "servers": self.servers,
            "shares": self.shares,
            "share_aces": self.share_aces,
            "directories": self.directories,
            "ntfs_aces": self.ntfs_aces,
            "scan_runs": self.scan_runs,
        }


@dataclass(frozen=True, slots=True)
class OperationsReport:
    """Everything the collector status page renders, assembled once."""

    status: CollectionStatus
    scopes: tuple[ScopeOperations, ...]
    counts: ObjectCounts
    errors: tuple[ErrorGroup, ...]

    @property
    def notes(self) -> tuple[str, ...]:
        """Every scope's note, in scope order. Empty when nothing needs attention."""
        return tuple(scope.note for scope in self.scopes if scope.note is not None)

    @property
    def never_succeeded(self) -> tuple[ScopeOperations, ...]:
        """Scopes with no successful run at all — the darkest parts of the estate."""
        return tuple(scope for scope in self.scopes if not scope.has_ever_succeeded)

    @property
    def total_errors(self) -> int:
        return sum(group.count for group in self.errors)


def _key(collector: str, target: str | None) -> tuple[str, str]:
    """The scope identity. ``target`` is coalesced because SQL groups it that way too."""
    return (collector, target or "")


def _moment(value: dt.datetime | None) -> str:
    return "an unknown time" if value is None else value.strftime("%Y-%m-%d %H:%M UTC")


def assemble_operations(
    *,
    status: CollectionStatus,
    latest: Iterable[RunOutcome],
    successes: Iterable[RunOutcome],
    failures: Iterable[RunOutcome],
    counts: ObjectCounts,
    errors: Iterable[ErrorGroup],
) -> OperationsReport:
    """Pair each scope's latest run with the last success and the last failure of that scope.

    ``successes`` and ``failures`` are each already reduced to one row per scope by the
    query; a scope absent from either simply has none. A success or failure naming a scope
    that has no latest run cannot happen — the latest run of a scope that has a successful
    run is at worst that run — and is dropped rather than invented, because a report with a
    scope in it that no run produced would be a report of something that did not happen.
    """
    by_scope = {_key(run.collector, run.target): run for run in latest}
    success_by_scope = {_key(run.collector, run.target): run for run in successes}
    failure_by_scope = {_key(run.collector, run.target): run for run in failures}

    scopes = tuple(
        ScopeOperations(
            collector=run.collector,
            target=run.target,
            latest=run,
            last_success=success_by_scope.get(key),
            last_failure=failure_by_scope.get(key),
        )
        for key, run in sorted(by_scope.items())
    )
    return OperationsReport(
        status=status,
        scopes=scopes,
        counts=counts,
        errors=tuple(sorted(errors, key=lambda group: (-group.count, group.code))),
    )
